### DISCUSSION TITLE
WebGL roadmap

### DISCUSSION THREAD

**Discussion by ianthomas23** [bokeh/bokeh#11369] [General]
Now that we have some (the most commonly used) functionality of `line`, `marker` and `rect` glyphs working using WebGL and have demonstrated that the approach is worthwhile from a performance perspective, here are plans looking forward for the WebGL backend.  Items are in priority order at the top of the list but not necessarily so at the bottom of the list.

1. ~Separate javascript bundle (issue #11053, fixed by #11371).~
2. Minifying shaders.
3. Fix known problems (remaining items of issues #11050, #11051 and #11052), in particular line dashes.
4. Support `scatter` with multiple different types of marker (done in PR #11840).
5. ~~Unify `marker` and `rect`, as `rect` is just a square `marker` with a 2D size expressed in data not screen units~~ (done in PR #12040).
6. Extend to cover other glyphs that are easy to implement (`line` to `multi_line`, `rect` to `quad` and `vbar`, etc).
7. Support arbitrary area glyphs, i.e. ones that may be concave and/or contain holes, e.g. `multi_polygon`s.  This needs tesselation of areas into convex polygons or triangles, which will requires an extra JavaScript library or perhaps a Wasm experiment.
8. Investigate sending untransformed data and transform to shaders rather than transformed data.  This will significantly reduce the amount of data sent to the GPU for some use cases which will improve performance.  Easy for linear axes, harder for log, categorical and datetime axes.
9. Images.
10. Specialised curves, e.g. `bezier`.

If anyone has different ideas for priorities here, now would be a good time to mention it!

(Edited to strikethrough completed items).

**Comment by bryevdv** [comment]
@ianthomas23 I think this is a great list and order. The only thing I'd mention in addition is that (IMO) looking for any opportunities to converge the webgl uniform/vector notions with the Bokeh "dataspec" notion as much as possible over time would also be good as a background goal.

**Reply by ianthomas23** [reply]
OK, I will bear that in mind.

**Comment by mattpap** [comment]
> Investigate sending untransformed data and transform to shaders rather than transformed data. This will significantly reduce the amount of data sent to the GPU for some use cases which will improve performance. Easy for linear axes, harder for log, categorical and datetime axes.

The original webgl backend used to work like this, though with a lot of limitations and unreliably. We decided to simplify this and let webgl focus only on rendering and leave any high-level processing to the data processing pipeline. If you can do better, that's great. However, keep in mind the maintenance burden and extensibility of existing capabilities when having effectively two implementations.
